ci: shard aio example e2e tests

petebacondarwin opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- ~~[ ] Tests for the changes have been added (for bug fixes / features)~~
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/angular/angular/issues/18230

The AIO example e2e tests take too long to run, and so are currently left as an optional build.

## What is the new behavior?

The AIO example e2e tests are split into 2 shards and are now compulsory for a successful build.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

mary-poppins commented:
You can preview a98f9eb at https://pr18238-a98f9eb.ngbuilds.io/.

mary-poppins commented:
You can preview 4321a39 at https://pr18238-4321a39.ngbuilds.io/.

mary-poppins commented:
You can preview 723144f at https://pr18238-723144f.ngbuilds.io/.

mary-poppins commented:
You can preview fd86be8 at https://pr18238-fd86be8.ngbuilds.io/.

mary-poppins commented:
You can preview f21cc6b at https://pr18238-f21cc6b.ngbuilds.io/.

mary-poppins commented:
You can preview 32faea8 at https://pr18238-32faea8.ngbuilds.io/.

petebacondarwin committed:
docs(aio): remove generated styles.css file

This file should have been ignored as it is created
during the build of the example

petebacondarwin committed:
ci(aio): test the example e2e files using local build of Angular

petebacondarwin committed:
docs(aio): fix deprecated protractor API usage

`browser.getLocationAbsUrl()` is deprecated.
We should use `browser.getCurrentUrl()` instead.

Foxandxss commented:
The changes are really nice Pete. Perhaps a PR with too many different changes tho.

Good tests too, you could test the zipper and plunkers 😂

petebacondarwin committed:
build(aio): ignore generated aot files

Assets such as images and data which are generated
by the aot build were not being ignored.

mary-poppins commented:
You can preview 28f9630 at https://pr18238-28f9630.ngbuilds.io/.

petebacondarwin committed:
build: short-circuit build for AIO tasks

mary-poppins commented:
You can preview be51d10 at https://pr18238-be51d10.ngbuilds.io/.

petebacondarwin committed:
docs(aio): delay ngUpgrade e2e test to avoid flakes

mary-poppins commented:
You can preview 240ee33 at https://pr18238-240ee33.ngbuilds.io/.

petebacondarwin committed:
ci: short-circuit npm install for aio builds that use yarn only

mary-poppins commented:
You can preview 7a14e21 at https://pr18238-7a14e21.ngbuilds.io/.

mary-poppins commented:
You can preview a272b25 at https://pr18238-a272b25.ngbuilds.io/.

mary-poppins commented:
You can preview e810e9f at https://pr18238-e810e9f.ngbuilds.io/.

petebacondarwin commented:
Except for the random `bazel` errors this passes Travis and brings the E2E tests for AIO down from 47 mins to 33 mins.

IgorMinar commented on aio/content/examples/upgrade-phonecat-2-hybrid/e2e-spec.ts:
```diff
@@ -9,7 +9,8 @@ describe('PhoneCat Application', function() {
   it('should redirect `index.html` to `index.html#!/phones', function() {
     browser.get('index.html');
-    expect(browser.getLocationAbsUrl()).toBe('/phones');
+    browser.sleep(1000); // Not sure why this is needed but it is. The route change works fine.
```
:-(

IgorMinar commented on aio/tools/examples/run-example-e2e.js:
```diff
@@ -22,15 +22,24 @@ const IGNORED_EXAMPLES = [
+ *
+ *  --shard to shard the specs into groups to allow you to run them in parallel
+ *    e.g. --shard=1/2 // every other spec: 1, 3, 5, etc
+ *    e.g. --shard=0/2 // every other spec: 0, 2, 4, etc
```
maybe swap the 0/2 and 1/2 lines to make the ordering right? also consider using odd/even terminology

petebacondarwin replied:
I avoided even/odd terminology because I had a feeling we might switch to 3 shards some point soon.

petebacondarwin replied:
But I will go with it.

IgorMinar commented on scripts/ci/test.sh:
```diff
@@ -44,7 +44,7 @@ case ${CI_MODE} in
     ;;
   aio_e2e)
-    ${thisDir}/test-aio-e2e.sh
+    ${thisDir}/test-aio-e2e.sh $AIO_SHARD
```
I would move this env variable into test-aio-e2e.sh to remove the level of indirection which is not helpful here.

petebacondarwin replied:
Right

IgorMinar reviewed (approved):
This looks quite good!

gkalpak commented on aio/package.json:
```diff
@@ -33,8 +33,9 @@
     "serve-and-sync": "concurrently --kill-others \"yarn docs-watch\" \"yarn start\"",
     "~~update-webdriver": "webdriver-manager update --standalone false --gecko false",
-    "boilerplate:add": "node ./tools/examples/add-example-boilerplate add",
-    "boilerplate:remove": "node ./tools/examples/add-example-boilerplate remove",
```
Remove `add-example-boilerplate.js`.

gkalpak commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -27,17 +27,45 @@ const BOILERPLATE_TEST_PATHS = [
+  'platform-browser-dynamic',
+  'platform-server',
+  'router',
+  'upgrade',
```
Oh, yet another list of packages :smiley:

gkalpak commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -67,7 +95,9 @@ class ExampleBoilerPlate {
       .usage('$0 <cmd> [args]')
-      .command('add', 'add the boilerplate to each example', argv => this.add(argv.local))
+      .command('add [--local]', 'add the boilerplate to each example',
+              { local: { describe: 'Use the locally built Angular libraries, rather than ones from  npm.' } },
```
Double space before `npm` 😱

gkalpak commented on aio/tools/examples/example-boilerplate.spec.js:
```diff
@@ -76,6 +87,23 @@ describe('example-boilerplate tool', () => {
+      spyOn(fs, 'ensureSymlinkSync');
+    });
+
+    it('should remove the shared node_modules folder', () => {
```
The description is wrong (or something is missing).

petebacondarwin replied:
Right, I'll tweak.

petebacondarwin committed:
ci: shard the aio example e2e tests

petebacondarwin committed:
build(aio): support overriding the Angular packages in examples with locally built ones

petebacondarwin committed:
build(aio): refactor and test the example-boilerplate tool

petebacondarwin committed:
ci(aio): support sharding of example e2e tests

mary-poppins commented:
You can preview a98f9eb at https://pr18238-a98f9eb.ngbuilds.io/.
You can preview e51b7eb at https://pr18238-e51b7eb.ngbuilds.io/.
You can preview 4321a39 at https://pr18238-4321a39.ngbuilds.io/.
You can preview 723144f at https://pr18238-723144f.ngbuilds.io/.
You can preview fd86be8 at https://pr18238-fd86be8.ngbuilds.io/.
You can preview f21cc6b at https://pr18238-f21cc6b.ngbuilds.io/.
You can preview 32faea8 at https://pr18238-32faea8.ngbuilds.io/.
You can preview 28f9630 at https://pr18238-28f9630.ngbuilds.io/.
You can preview be51d10 at https://pr18238-be51d10.ngbuilds.io/.
You can preview 240ee33 at https://pr18238-240ee33.ngbuilds.io/.
You can preview 7a14e21 at https://pr18238-7a14e21.ngbuilds.io/.
You can preview a272b25 at https://pr18238-a272b25.ngbuilds.io/.
You can preview e810e9f at https://pr18238-e810e9f.ngbuilds.io/.

mary-poppins commented:
You can preview e91dbdd at https://pr18238-e91dbdd.ngbuilds.io/.

petebacondarwin commented:
The only failure is the known core unit test problem, which is not related to this PR.
